### DISCUSSION TITLE
analogRead seems not to work in homespan project

### DISCUSSION THREAD

**Discussion by zhekch** [HomeSpan/HomeSpan#347] [Q&A] [answered]
When I try to read analog pin, it just shows a zero, however when I make a new project that also read analog input, it works. I've added a [video](https://drive.google.com/file/d/1wIoG13gMyR7ikhu_3rHkQPjki_S-Voro/view?usp=sharing) that shows the problem

```
#include <DHT.h>

#define DHTTYPE DHT11
#define DHTPIN 13
int CO2;
int aQ;
DHT dht(DHTPIN, DHTTYPE);

//CUSTOM_SERV(MotionSensor, ad0663c4-8520-4039-a45e-d89c2ffaf70f);
//CUSTOM_CHAR(motionDetected, ad0663c4-8520-4039-a45e-d89c2ffaf70f,PR+EV, BOOLEAN, 0, 0, 1, false);

struct DEV_TempSensor : Service::TemperatureSensor {
  SpanCharacteristic *temp;
  DEV_TempSensor() : Service::TemperatureSensor() {
    dht.begin();
    temp = new Characteristic::CurrentTemperature(-10.0);
    // expand the range from the HAP default of 0-100 to -50 to 100 to allow for negative temperatures
    temp->setRange(-50, 100);

    // initialization message
    Serial.print("Configuring Temperature Sensor");
    Serial.print("\n");

  } // end constructor

  void loop() {
    // the temperature refreshes every 10 seconds by the elapsed time
    if (temp->timeVal() > 10000) {
      // read temperature from sensor dht22
      float temperature = dht.readTemperature();
      // set the new temperature; this generates an Event Notification and also resets the elapsed time
      temp->setVal(temperature);

      Serial.print("\nTemperature Update: ");
      Serial.print(temperature);
      Serial.print(" ; ");
    }
  }
};






struct DEV_HumSensor : Service::HumiditySensor {

  SpanCharacteristic *hum;

  // constructor() method
  DEV_HumSensor() : Service::HumiditySensor() {


    dht.begin();

    // instantiate the Current Temperature Characteristic
    hum = new Characteristic::CurrentRelativeHumidity(50);


    // initialization message
    Serial.print("Configuring Humidity Sensor");
    Serial.print("\n");

  }

  void loop() {

    // the humidity refreshes every 10 seconds by the elapsed time
    if (hum->timeVal() > 10000) {

      float Humi = dht.readHumidity();
      hum->setVal(Humi);
      Serial.print("Humidity Update: ");
      Serial.print(Humi);
      Serial.print(" ; ");
    }
  }
};


struct DEV_AirQualitySensor : Service::AirQualitySensor {     // A standalone Air Quality sensor

  SpanCharacteristic *airQuality;
  SpanCharacteristic *CarbonDioxideDetected;
  SpanCharacteristic *CarbonDioxideLevel;

  DEV_AirQualitySensor() : Service::AirQualitySensor() {      // constructor() method

    airQuality = new Characteristic::AirQuality(1);
    CarbonDioxideDetected = new Characteristic::CarbonDioxideDetected(false);
    CarbonDioxideLevel = new Characteristic::CarbonDioxideLevel(500);

    Serial.print("Configuring Air Quality Sensor");
    Serial.print("\n");
  }


  void loop() {
    
    if (airQuality->timeVal() > 5000) {
      aQ = map(analogRead(4),0,1023,1,5);
      airQuality->setVal(aQ);
      CO2 = analogRead(4);                //get compensated value
      if (CO2 > 800) {
        CarbonDioxideDetected->setVal(true);
      }
      else {
        CarbonDioxideDetected->setVal(false);
      }
      if ((CO2 >= 1300 or CO2 <= 500)) {
        CarbonDioxideLevel->setVal(0);
      } else {

        CarbonDioxideLevel->setVal(CO2);
      }
      Serial.print("Air Quality Update: ");
      Serial.print(aQ);
      Serial.print("; ");
      Serial.print("CO2: ");
      Serial.print(analogRead(4));
      Serial.print(", ");
      
    }
  }
};

```

**Comment by HomeSpan** [comment]
Can you provide your code.

**Reply by zhekch** [reply]
Also it was working for me, but like yesterday I left it over the night, and when I looked at Serial output, the number was out of range, because analog input stopped working for some reason

**Comment by HomeSpan** [comment]
If it's working, and then after a while it stops, that suggests something unrelated to HomeSpan.  I don't know which board you are using, but you may want to double-check that pin 4 is not reserved for use by anything else.

**Reply by zhekch** [reply]
I still think that it is somehow related to homespan, cause no matter which pin I'm using, analogRead and digitalRead dont work at all, but I'll try to do something with it, like creating new files and rewriting code. Also I use ESP32 dev module.

**Comment by HomeSpan** [comment]
digitalRead doesn't work either?  Are you setting input mode with pinMode?  Also, I've only tested HomeSpan up to version 2.0.3 of the Arduino-ESP32 board manager - are you using 2.0.4 (I will be testing that shortly - already found one issue Espressif introduced).

**Reply by zhekch** [reply]
Sorry, digitalRead does work, however pinMode doesn't help, Serial monitor still shows 0 for analogRead. But one thing that I noticed is that the library for DHT11 works completely fine, but as far as I know the library doesn't use analogRead to get data from the sensor

**Comment by HomeSpan** [comment]
I have been able to reproduce - will try to debug.

**Comment by HomeSpan** [comment, marked as answer]
I found the source of the problem.  It's not with HomeSpan but rather the problem is incompatibility with analog pins using ADC2 and WiFi.  See [here](https://community.blynk.cc/t/interesting-esp32-issue-cant-use-analogread-with-wifi-and-or-esp-wifimanager-library/49130) for an explanation.  Solution is to only use analog pins associated with ADC1, which are pins 32-39.  I tested using analogRead with pin36 inside a HomeSpan sketch and it works fine.

**Reply by zhekch** [reply]
Thanks, I tried pin 34 and it seems that the problem has been solved, but I guess you meant ADC1 instead of ADC2. Anyway thanks once again for the solution
